CMSE calling conventions (#3884)
rust-lang/rfcs: text/0000-cmse-calling-conventions.md

- Feature Name:  `cmse_nonsecure_entry` and `abi_cmse_nonsecure_call`
- Start Date: 2025-11-24
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/3884)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> Support for the `cmse-nonsecure-entry` and `cmse-nonsecure-call` calling conventions on   `thumbv8` targets, and a lint preventing (partially) uninitialized values from crossing the security boundary.

Reviewer: ```suggestion
Support for the `cmse-nonsecure-entry` and `cmse-nonsecure-call` calling conventions on   Armv8-M (`thumbv8m*`) targets, and a lint preventing (partially) uninitialized values from crossing the security boundary.
```

Functions that use these ABIs must be reviewed carefully because they mark where secure data might be leaked. This is analogous to `unsafe` limiting  where UB might be introduced in a program. The calling conventions automatically handle the clearing of registers before the secure boundary is crossed, so that a malicious non-secure application cannot read lingering secure data.

Without compiler support it is much harder to know where to focus review effort, and every call that crosses the secure boundary requires inline assembly, which is inconvenient and error-prone.

A specific use case is encapsulating C APIs. Providing a C interface is still the standard way for a hardware vendor to provide access to system components. Libraries for networking (LTE, Bluetooth) are notorious for their bugs. Running such code in non-secure mode significantly reduces the risk of bugs leaking secure information.

Trustzone is growing in availability and use. More and more of the new medium and large ARM microcontrollers have support. Large industry players have requested Rust support for Trustzone.
# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> The cmse calling conventions are part of the *Cortex-M Security Extension* that are available on thumbv8 systems. They are used together with Trustzone (hardware isolation) to create more secure embedded applications.

Reviewer: The architecture is called Armv8-M. The targets that support this architecture happen to begin `thumbv8`, but so would a target for Thumb-mode on Armv8-R.

The `cmse-nonsecure-call` calling convention is used in the other direction, when the secure executable wants to call into the non-secure executable. This calling convention can only occur on function pointers, not on definitions or extern blocks. The secure executable can acquire a non-secure function pointer via shared memory, or a non-secure callback can be passed to an entry function.

Both calling conventions are based on the platform's C calling convention, but will not use the stack to pass arguments or the return value. In practice that means that the arguments must fit in the 4 available argument registers, and the return value must fit in a single 32-bit register, or be abi-compatible with a 64-bit integer or float. The compiler checks that the signature is valid.
# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Arm defines the toolchain requirements in  [ARMv8-M Security Extensions: Requirements on Development Tools - Engineering Specification](https://developer.arm.com/documentation/ecm0359818/latest/), but of course this specification needs to be interpreted in a Rust context.
## ABI Details

> The `cmse-nonsecure-call` and `cmse-nonsecure-entry` ABIs are only accepted on `thumbv8m-*` targets, . On all other targets their use emits an invalid ABI error.

Reviewer: ```suggestion
The `cmse-nonsecure-call` and `cmse-nonsecure-entry` ABIs are only accepted on Armv8-M targets (currently `thumbv8m.base-*-eabi`, `thumbv8m.main-*-eabi{,hf}`). On all other targets their use emits an invalid ABI error.
```

The foundation of the cmse ABIs is the platform's standard AAPCS calling convention. On `thumbv8m` targets `extern "aapcs"` is the default C ABI and equivalent to `extern "C"`.

The `cmse-nonsecure-call` ABI can only be used on function pointers. Using it in for a function definition or extern block emits an error. It is invalid to cast to or from `extern "aapcs"`.

> The `cmse-nonsecure-entry` ABI is allowed on function definitions, extern blocks and function pointers. It is sound and valid (in some cases even encouraged) to cast such a function to `extern "aapcs"`. Calling the function is valid and will behave as expected in both the secure and non-secure applications. Casting from `extern "aapcs"` to `extern "C"` is invalid.

Reviewer 2: Is this issue resolved now?

## ABI Details

The `cmse-nonsecure-call` and `cmse-nonsecure-entry` ABIs are only accepted on Armv8-M targets (currently `thumbv8m.base-*-eabi`, `thumbv8m.main-*-eabi{,hf}`). On all other targets their use emits an invalid ABI error.

The foundation of the cmse ABIs is the platform's standard AAPCS calling convention. On `thumbv8m` targets `extern "aapcs"` is the default C ABI and equivalent to `extern "C"`.

The `cmse-nonsecure-call` ABI can only be used on function pointers. Using it in for a function definition or extern block emits an error. It is invalid to cast to or from `extern "aapcs"`.

The `cmse-nonsecure-entry` ABI is allowed on function definitions, extern blocks and function pointers. It is sound and valid (in some cases even encouraged) to convert such a function to `extern "aapcs"`, particularly in the non-secure application.

```rust
#[unsafe(no_mangle)]
extern "cmse-nonsecure-entry" fn foo(x: i32) -> i32 { x + x }

unsafe extern "aapcs" {
    #[link_name = "foo"]
    safe fn bar(x: i32) -> i32;
}
```

> Calling the `bar` function is valid and will behave like `foo` in both the secure and non-secure applications. A transmute between `extern "cmse-nonsecure-entry"` and `extern "aapcs"` is similarly valid. Converting in the other direction, from `extern "aapcs"` to `extern "cmse-nonsecure-entry"`, is invalid.

Reviewer 3: ```suggestion
Calling the `bar` function is valid and will behave like `foo` in both the secure and non-secure applications. A transmute from `extern "cmse-nonsecure-entry"` to `extern "aapcs"` is similarly valid. Converting in the other direction, from `extern "aapcs"` to `extern "cmse-nonsecure-entry"`, is invalid.
```

using `between` kinda means both directions to me, imo it's better to use `from .. to ..`

### Argument passing

The main technical limitation of the cmse ABIs versus plain AAPCS is that the cmse ABIs cannot use the stack for passing function arguments or return values. That leaves only the 4 standard registers to pass arguments, and only supports 1 register worth of return value, unless the return type is ABI-compatible with a 64-bit scalar, which is supported.

```rust
// Valid
type T0 = extern "cmse-nonsecure-call" fn(_: i32, _: i32, _: i32, _: i32) -> i32;
type T1 = extern "cmse-nonsecure-call" fn(_: i64, _: i64) -> i64;

#[repr(transparent)] struct U64(u64);
type T2 = extern "cmse-nonsecure-call" fn() -> U64;

// Invalid: too many argument registers used
type T3 = extern "cmse-nonsecure-call" fn(_: i64, _: u8, _: u8, _: u8) -> i64;

// Invalid: return type too large
type T4 = extern "cmse-nonsecure-call" fn() -> i128;

// Invalid: return type does not fit in one register, and is not abi-compatible with a 64-bit scalar
#[repr(C)] struct WrappedI64(i64);
type T5 = extern "cmse-nonsecure-call" fn(_: i64, _: i64) -> WrappedI64;
```

> An error is emitted when the program contains a signature that violates the calling convention's constraints:

Reviewer 4: `BackendRepr` is an internal implementation detail of rustc and prone to changes. It seems risky to get ourselves into a corner where changing unspecified layout computation details is a breaking change.

> An error is emitted when the program contains a signature that violates the calling convention's constraints:

Reviewer 4: > We test for the case you mention

That is not the case I mentioned. I suggested to remove `repr(transparent)`, but I didn't suggest adding `repr(C)`. (That will probably trigger an FFI lint, but since we are talking about hard errors here, whether or not there is a lint is largely irrelevant.)

Also, the fact that `repr(C)` inhibits `BackendRepr::Scalar` is a limitation we'd like to lift (https://github.com/rust-lang/rust/issues/119183). We should not add new logic that relies on how `repr(C)` and `BackendRepr` correlate.

We'd like to wait with adding further lints until we see more usage of Trustzone, so that we can design a lint that covers all practical cases of too-strong assumptions.
## An "initialize padding" attribute

The current lint for partially uninitialized values crossing the security boundary does not have a proper workaround: the advice is to just not send such values over the secure boundary, and essentially treat the warning as an error.

A suggestion that was floated is to provide some mechanism to ensure that a value is fully initialized, e.g. by zeroing any potentially uninitialized parts.

One potential method is to extend the`repr` attribute with an option that adds fields where padding is needed internally. These user-hidden padding fields would be zeroed upon creation.
```rust
#[repr(C, align(8), initialized)]
struct Foo {
	a: u8,
	// implicit _padding0: [u8; 1],
	b: u16,
	// implicit _padding1: [u8; 4],
}
```

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Author: The padding is overwritten only in simple cases. E.g. as soon as unions are involved it all gets more complicated and `clang` does not overwrite the padding https://godbolt.org/z/xebMTd67f. So for padding I think the lint is fine, it's totally workable for engineers taking on a cmse project.

If it does turn out to be cumbersome we can implement the zeroing in some cases, and stop linting for them.

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Reviewer 5: Well yes, in that latter example it's impossible to know which variant is meant to be active, so naturally clang can't zero and has to warn. In fact, the union doesn't actually have padding at all. With the equivalent Rust union, you could do:

```rust
#[repr(C)]
#[derive(Clone, Copy)]
struct Padding1 {
    x: u16,
    z: u8,
    y: u32,
}

#[repr(C)]
#[derive(Clone, Copy)]
struct Padding2 {
    y: u32,
    z: u8,
    x: u16,
}

#[repr(C)]
union PaddingSomewhere {
    a: Padding1,
    b: Padding2,
}

fn main() {
    let mut foo = PaddingSomewhere {
        a: Padding1 { x: 0, z: 0, y: 42 },
    };
    foo.b.y = 42;
    // all bytes of `foo` are now initialized, and all must be preserved across a function call
}
```

The proper semantic, which matches Clang, is to zero padding bytes, while warning for bytes whose initialization state is unknown but must be preserved (like `MaybeUninit`, or the union above).

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Author: @RalfJung do we have, or do you see a somewhat convenient way to implement, a copy that guarantees that padding (at a compile-time known position) is not copied? I'd really prefer not to do something ad-hoc here. 

As discussed we can't know at compile time what bytes of an `enum`/`union` are initialized, so those should still trigger the lint if the type may be partially uninitialized.

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Reviewer 5: `enum`s could be made to work, you could match on the variant at runtime as @RalfJung suggested. (Whether that is worth the effort to implement is another matter…)

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Reviewer 4: I'm not quite sure what you mean, but I think the answer is no. Padding is by definition left undefined in the destination precisely so that the compiler can choose to copy or not copy.

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Author: Well really I'm just looking for a way to actually implement this without creating something really custom that becomes a maintenance burden because it's hooked into all function calls.

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Reviewer 5: FWIW, "zero all the padding in this value" is a primitive that could be useful beyond CMSE. E.g. for analogous features on other targets, but also potentially serialization, `AtomicCell`, etc.

> This feature still has many open design questions. We don't think such an attribute is required for practical Trustzone development, so we defer it for now.

Author: As of https://github.com/rust-lang/rust/pull/157397 the guaranteed (so, variant-independent) padding bytes are now cleared before crossing the secure boundary.

### No C-variadics (currently)

Currently both ABIs disallow the use of c-variadics. For `cmse-nonsecure-entry`, the toolchain actually does not support c-variadic signatures (likely because of how they interact with shim that switches to secure mode, though the specification does not say that explicitly).

- clang rejects c-variadic entry functions: https://godbolt.org/z/MaPjzGcE1
- but accepts c-variadic nonsecure calls: https://godbolt.org/z/5rdK58ar4

For `cmse-nonsecure-call`, we may support and stabilize c-variadics at some point in the future.

## No tail calls

Neither cmse ABI can tail call another function, per the LLVM source:

> For both the non-secure calls and the returns from a CMSE entry function, the function needs to do some extra work after the call, or before the return, respectively, thus it cannot end with a tail call

The unstable implementation of guaranteed tail calls in rust requires the caller and callee to have the same ABI. That means that calls to `cmse-nonsecure-call` are never eligible for a tail call (there are no definitions with this ABI). For tail calls to a `cmse-nonsecure-entry` function we emit an explicit error.

> Functions with the `extern "cmse-nonsecure-entry"` ABI may themselves be tail-called, though this is only possible when the function is first cast to `extern "C"` to satisfy the restriction that caller and callee have the same ABI.

Reviewer 4: "For tail calls to a `cmse-nonsecure-entry` function we emit an explicit error. [...] Functions with the `extern "cmse-nonsecure-entry"` ABI may themselves be tail-called"

The 2nd sentence is saying the opposite of the first?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3884/84057dbb428b770e1d9ff5999cb01d22d3fe44a8..da4c8672ddaf8fd370bddb688dc15ab40bbd1e7d)*

Author: Clang is a bit less restrictive when it comes to the conditions on a tail call: it recognizes that `cmse_nonsecure_entry` is really just `aapcs` with some retrictions, and so a plain function can tail call a `cmse_nonsecure_entry` function if the signatures otherwise match. So this works (I'll clarify the text):

https://godbolt.org/z/P3ToKTW94

```c
__attribute__((noinline))
__attribute__((cmse_nonsecure_entry))
int callee(int x)
{
    return x + 1;
}

// NOTE: no cmse attribute, this is a normal C function. 
int caller(int x)
{
    __attribute__((musttail))
    return callee(x);
}
```

### Support for `const fn`

No special support for calling cmse functions is needed.

Evaluating entry functions during constant evaluation is valid. The context switch from non-secure to secure mode is handled by the shim that switches to secure mode, which is not visible to rust code. Clearing of registers is not relevant for constant evaluation.

The `cmse-nonsecure-call` calling convention can only be used on function pointers, which cannot be evaluated during constant evaluation.

Miri is not a register machine, so the clearing of registers is not relevant. The context switching also does not need to be considered, because a Miri input program cannot use FFI and therefore cannot cross the secure boundary. Any attempt to do so would rely on a transmute or similar and would for that reason be unsound.

> ## Preventing information leaking via padding bytes

Reviewer 4: It is a bit odd to me that we're building all this machinery to prevent leakage in padding of by-value arguments to these functions, but do nothing about leakage in padding of data that's behind a pointer. Is it just not possible to pass a pointer over this ABI?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3884/84057dbb428b770e1d9ff5999cb01d22d3fe44a8..da4c8672ddaf8fd370bddb688dc15ab40bbd1e7d)*

Reviewer 6: The pointer itself should be fine, but the thing it points to might live in secure memory. When it's dereferenced, the CPU will generate a secure access fault and the access is stopped.

So there's no danger there and to me it makes sense to be able to use opaque pointers for bookkeeping

Reviewer 7: That is not entirely true, right? After the secure call, the processor is in secure mode so it will happily dereference pointers to secure memory. The correct way is to first use the TrustZone-M instruction [`TT`](https://developer.arm.com/documentation/100690/0200/The-Test-Target-instruction) to test the target address range. That way, the secure world can guarantee that the non-secure world passed a pointer to non-secure memory.

Reviewer 6: What I wrote was in the context of going from secure to nonsecure.
When going the other way around, from nonsecure to secure, a pointer can be dangerous yes. But like I said, you may want opaque pointers or pointers into non-secure memory. It doesn't make sense to me to disallow pointers.

The memory config is done at runtime or linkertime, so I don't think the ABI could prevent misuse of pointers. So this is something where the implementer of the secure side needs to be vigilant.

The only thing the compiler could do is emit a warning maybe when a pointer is used?

Reviewer 4: Is there any situation where the secure and insecure part share memory, and the secure part needs to write some data to that shared memory in a way that carefully avoids leaking anything via the padding?

Reviewer 6: I see where you're going now. Yes, this happens. The most famous example is the secure side having the encryption keys and the nonsecure side calling a secure function to decrypt or encrypt a slice of bytes. But you can easily imagine this kind of stuff happen with more complex datatypes than a byte slice

Reviewer 4: Then it sounds like we'd want a facility to "clean" an arbitrary type's padding, instead of having this magic be only available for CMSE calls.

Though thinking about it a bit more -- even if we did, function calls incur their own copies so we'd still need to also do something special on those calls, there's no way for the user to do that themselves. So this should probably just be mentioned as a future possibility.

> Values that cross the secure boundary may have stale secure data in their padding bytes. Padding can be either variant-dependent or variant-independent. 

Reviewer 8: I see in the (I think latest?) definition from Arm that what I think is f16 is passed with indeterminate bits:

> If the argument is a Half-precision Floating Point Type its size is set to 4 bytes as if it had been copied to the least significant bits of a 32-bit register and the remaining bits filled with unspecified values.

(https://github.com/ARM-software/abi-aa/blob/daa7a94ca55973736c0e434a67a6e4bbcd35d7fa/aapcs32/aapcs32.rst#65parameter-passing)

Does that imply we need to lint on such a type, even though from a Rust perspective it has no padding? Or does llvm for cmse calling conventions actually zero the extra bytes?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3884/84057dbb428b770e1d9ff5999cb01d22d3fe44a8..da4c8672ddaf8fd370bddb688dc15ab40bbd1e7d)*

Reviewer 4: This might happen with other types as well, e.g. u16 or u8 or bool.

Author: > This might happen with other types as well, e.g. u16 or u8 or bool.

No because

> If the argument is an integral Fundamental Data Type that is smaller than a word, then it is zero- or sign-extended to a full word and its size is set to 4 bytes.

---

`f16` is fine, a `uxth` "unsigned extend halfword" zeroes the upper bits:

- entry https://godbolt.org/z/4T8qTvEs9
- call https://godbolt.org/z/9GMWxsee8

---

composite types similarly zero extend, e.g. https://godbolt.org/z/7W7TM4jsr

Author: I'll add an assembly test to ensure that doesn't regress

A `cmse-nonsecure-call` function call will emit a warning when any of its arguments has a partially uninitialized type, and a `cmse-nonsecure-entry` function warns at any (implicit) return when the return type may be partially uninitialized.

Ultimately guaranteeing the security properties of the system is up to the programmer, but warning on types with potentially uninitialized memory is a helpful signal that the compiler can provide.

#### Implementation

https://github.com/rust-lang/rust/pull/147697

Much like for variant-independent padding, the implementation walks over a type, but instead of using the union to combine the data bytes of the variants, the intersection is used. That gives a map of bytes that are data in all variants. 

We then take the difference with the map of bytes that are data in any variant, and are left with the set of ranges that is sometimes, but not always, data: the variant-dependent padding.

If this set is non-empty, the lint is triggered.

> ## Background

Reviewer 8: Can you say more here about the common/expected usage in Rust code? For example, if we did similar to variadics and restricted to an extremely limited set of types, would that be meaningfully painful for users? I saw some mention in comments that ZSTs being present in the argument list (and ignored) is meaningfully useful, is that the limit?

In short, I'm wondering if we can sidestep some of the stability questions by only allowing types we currently guarantee ABI compatibility for with the core platform types (e.g., u32, f32, ...) and leave the rest to users to wrap to start.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3884/84057dbb428b770e1d9ff5999cb01d22d3fe44a8..da4c8672ddaf8fd370bddb688dc15ab40bbd1e7d)*

Author: Only primitives would be extremely restrictive here. As mentioned ZSTs are used extensively in the rust embedded ecosystem, so those should be allowed.

With c-variadic part of the problem is that LLVM just doesn't support passing arbitrary types for all platforms: we lock down what types are allowed to the set that we know works. 

The real problem here is leaking details about `repr(Rust)`, i.e. when the layout algorithm changes a program might stop compiling. So a restriction that prevents such breakage is to say that a type that is (or recursively contains) `repr(Rust)` should be disallowed, perhaps with an exception for ZSTs. E.g. `repr(C)` is fine because the layout algorithm is stable.

I suspect that is too limiting, because most types in the rust embedded ecosystem are `repr(Rust)`. Maybe actual users can weigh in on what such a restriction would mean in practice (cc @diondokter).

Reviewer 6: Yes, that seems too restrictive to me.
A lot of usecases will be fairly simple. But you'd still want to be able to pass common rust things like Options, Results, Enums, slices, etc.

Also something like a MaybeUninit wouldn't work if unions weren't allowed.

Reviewer 4: > Also something like a MaybeUninit wouldn't work if unions weren't allowed.

I would think that *shouldn't* work because who knows which data is hiding in that MaybeUninit.

Reviewer 6: Maybe not from secure to nonsecure. But from nonsecure to secure it makes sense to want to use it

Author: The idea of the lint is that you can `allow` it if you actually know that it's fine. It's an FFI boundary, it's all unsafe anyway, we're just trying to prevent accidental leakage.

Reviewer 4: > Maybe not from secure to nonsecure. But from nonsecure to secure it makes sense to want to use it

In that direction I assume we don't do any padding clearing or so, none of this should be relevant?

Reviewer 6: For maybeuninit specifically yes. The other things I mentioned are relevant

Reviewer 4: I am confused. My assumption is that that the entire thread and the entire discussion about clearing padding is only relevant for secure-to-insecure calls/returns. So the insecure-to-secure part just seems entirely off-topic here.

Reviewer 5: > > Also something like a MaybeUninit wouldn't work if unions weren't allowed.
> 
> I would think that _shouldn't_ work because who knows which data is hiding in that MaybeUninit.

In general it is dubious, but maybe in someone's specific case they know they initialized it and it's fine

To ensure that the non-secure executable cannot read any lingering secret values from those registers, a call to a `cmse-nonsecure-call` function will clear all registers except those used to pass arguments.

A *non-secure function pointer*, i.e. a function pointer using the `cmse-nonsecure-call` calling convention, has its least significant bit (LSB) unset. Checking for whether this bit is set provides a way to test at runtime which security state is targeted by the function.

The secure executable can get its hands on a non-secure function pointer in two ways: the function address can be an argument to an entry function, or it can be in memory at a statically-known address.

# Drawbacks
[drawbacks]: #drawbacks

The usual reasons: this is a niche feature (although it is requested by large industry players) with a fair amount of complexity that must be maintained. However to be fair, these calling conventions have been in the compiler for around 5 years and so far the maintenance burden has been acceptable.

> # Rationale and alternatives

Reviewer 4: Another alternative worth mentioning is that unions in an argument type could just lead to a hard error. What's the reason for allowing unions at all? "clang allows it and just gives a warning"? If that's it then I think we can start with a hard error and weaken it to a deny-by-default lint if there is a usecase.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3884/84057dbb428b770e1d9ff5999cb01d22d3fe44a8..da4c8672ddaf8fd370bddb688dc15ab40bbd1e7d)*

}
```

Calling the `bar` function is valid and will behave like `foo` in both the secure and non-secure applications. A transmute from `extern "cmse-nonsecure-entry"` to `extern "aapcs"` is similarly valid. Converting in the other direction, from `extern "aapcs"` to `extern "cmse-nonsecure-entry"`, is invalid.

### Argument passing

The main technical limitation of the cmse ABIs versus plain AAPCS is that the cmse ABIs cannot use the stack for passing function arguments or return values. That leaves only the 4 standard registers to pass arguments, and only supports 1 register worth of return value, unless the return type is a (transparently wrapped) 64-bit scalar (`f64`, `i64` or `u64`), which is supported.

```rust
// Valid
type T0 = extern "cmse-nonsecure-call" fn(_: i32, _: i32, _: i32, _: i32) -> i32;
type T1 = extern "cmse-nonsecure-call" fn(_: i64, _: i64) -> i64;

#[repr(transparent)] struct U64(u64);
type T2 = extern "cmse-nonsecure-call" fn() -> U64;

// Invalid: too many argument registers used
type T3 = extern "cmse-nonsecure-call" fn(_: i64, _: u8, _: u8, _: u8) -> i64;

// Invalid: too many argument registers used (due to the alignment of `i64`).

> type T4 = extern "cmse-nonsecure-call" fn(_: i32, _: i64, _: ui32) -> i64;

Reviewer 7: There's a small typo in the type of the third argument (`u32` or `i32`)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3884/84057dbb428b770e1d9ff5999cb01d22d3fe44a8..39933a9f4b779f1c2792d392f29502f23e376bef)*

Author: ```suggestion
type T4 = extern "cmse-nonsecure-call" fn(_: i32, _: i64, _: u32) -> i64;
```
